File: examples/use_explorer_shell_windows.java
import com.jacob.com.ComThread;
import com.jacob.com.Dispatch;
import com.jacob.com.EnumVariant;
import com.jacob.com.Variant;

// javac -cp lib\jacob.jar use_explorer_shell_windows.java
// java -cp lib\jacob.jar;. -Djava.library.path=lib use_explorer_shell_windows
public class use_explorer_shell_windows {
	public static void main(String[] args) {
		ComThread.InitMTA();
		try {
			Dispatch shellWindows = new Dispatch("clsid:9BA05972-F6A8-11CF-A442-00A0C90A8F39");
			for (EnumVariant e = new EnumVariant(shellWindows); e.hasMoreElements(); ) {
				Variant fileVariant = (Variant) e.nextElement();
				Dispatch file = fileVariant.toDispatch();
				System.out.println(Dispatch.get(file, "Path").getString());
			}
		} finally {
			ComThread.Release();
		}
	}
}

File: examples/use_explorer_shell_windows.rb
# start me yup with jruby -I..\lib use_explorer_shell_windows.rb
# NB: make sure jruby script has java.library.path set, eg: _VM_OPTS=-Djava.library.path=lib;..\lib
require 'win32ole'

# These clsid came from shdocvw.dll type library (tlb/olb).
CLSID_ShellWindows = '{9BA05972-F6A8-11CF-A442-00A0C90A8F39}' # 'Shell.Application' coclass.

# Lets see which shell windows are opened.
shell = WIN32OLE.new(CLSID_ShellWindows)
shell.each do |window|
  puts window.Path
end

puts 'good bye...'

# XXX this hangs here because no one called ComThread.Release() on the Java Side!